helm: Add automatic StorageClass import/adoption to Rook Helm charts

dcaputo-harmoni opened the pull request:
## Problem Statement

When using Rook Helm charts in Infrastructure as Code (IaC) environments, users encounter StorageClass immutability conflicts that prevent seamless management of both new and existing Rook deployments.

### Current Issues:

1. **StorageClass Creation Conflicts**: If a StorageClass already exists (from previous manual deployment or Helm installation), subsequent Helm operations fail with:

   ```
   StorageClass.storage.k8s.io "ceph-block" is invalid: parameters: Forbidden: updates to parameters are forbidden
   ```

2. **Manual Workarounds Required**: Users must manually:

   - Set `storageClass.enabled: false` to avoid conflicts
   - Delete existing StorageClasses before Helm upgrades
   - Manage StorageClasses separately from pool configurations

3. **IaC Compatibility Problems**: The same Helm configuration cannot cleanly apply to both:

   - Fresh deployments (where StorageClasses need creation)
   - Existing deployments (where StorageClasses already exist)

4. **Configuration Verbosity**: Users must explicitly configure `storageClass` block even for minimal deployments, breaking the principle of sensible defaults.

### Example Failure Scenario:

**Step 1**: Initial Rook deployment using defaults (no `cephBlockPools` specified in Helm values)

- Helm creates default CephBlockPool with `size: 3`
- Helm creates default StorageClass "ceph-block"

**Step 2**: Later, user wants to modify pool configuration (reduce replica size for storage efficiency):

```yaml
# User adds this to Helm values to override default behavior
cephBlockPools:
  - name: ceph-blockpool # Successfully detects existing pool
    spec:
      replicated:
        size: 2 # Successfully updates pool spec
    storageClass:
      enabled: true
      name: ceph-block # FAILS: tries to patch existing StorageClass
```

**Result**:

- ✅ **CephBlockPool**: Helm detects existing pool and successfully updates `size: 3 → 2`
- ❌ **StorageClass**: Helm tries to patch existing "ceph-block" StorageClass → **Immutability error**

**Current Asymmetric Behavior**:

- Pool resources: Smart detection and merge ✅
- StorageClass resources: Patch attempts causing conflicts ❌

This forces users to add `storageClass.enabled: false` as a workaround, but then they lose Helm management of the StorageClass entirely.

## Solution

This PR implements **automatic StorageClass import/adoption** that eliminates conflicts and enables seamless IaC workflows.

### Key Changes:

1. **Automatic Default Merging**: Users can specify minimal configuration without explicit `storageClass` blocks:

   ```yaml
   cephBlockPools:
     - name: ceph-blockpool
       spec:
         replicated:
           size: 2
   # StorageClass defaults automatically applied: enabled=true, name="ceph-block"
   ```

2. **Intelligent StorageClass Management**:

   - **If StorageClass exists** → Silently adopt it (no YAML rendered, no Helm conflicts)
   - **If StorageClass doesn't exist** → Create it with sensible defaults
   - **No user intervention required**

3. **IaC-Friendly**: Same configuration works for both fresh and existing deployments without modifications.

### Technical Implementation:

- Uses Helm `lookup` function to detect existing StorageClasses during template rendering
- Prevents rendering StorageClass YAML when resource already exists
- Merges user configuration with sensible defaults using `mergeOverwrite`
- Applies to all storage types: CephBlockPool, CephFileSystem, CephObjectStore

## Benefits

### 🎯 **Seamless IaC Integration**

- Single Helm configuration works for new AND existing deployments
- No environment-specific workarounds or conditional logic needed
- Clean apply/upgrade cycles in Terraform, GitOps, and CI/CD pipelines

### 🚀 **Improved User Experience**

- Minimal configuration required - no verbose `storageClass` blocks for basic use cases
- Automatic conflict resolution - no more manual StorageClass deletion
- Backward compatible - existing configurations continue to work unchanged

### 🛡️ **Zero Breaking Changes**

- All existing parameters and overrides continue to work
- Users can still disable StorageClass creation with `storageClass.enabled: false`
- Custom StorageClass names and parameters remain fully supported

## Usage Examples

### Minimal Configuration (New Default Experience):

```yaml
cephBlockPools:
  - name: ceph-blockpool
    spec:
      replicated:
        size: 2
```

**Result**: Automatically creates "ceph-block" StorageClass if missing, adopts if existing.

### Custom Overrides (Still Supported):

```yaml
cephBlockPools:
  - name: ceph-blockpool
    spec:
      replicated:
        size: 2
    storageClass:
      name: "my-custom-storage-class"
      isDefault: false
      parameters:
        imageFeatures: "layering,exclusive-lock"
```

### Disable StorageClass (Still Supported):

```yaml
cephBlockPools:
  - name: ceph-blockpool
    spec:
      replicated:
        size: 2
    storageClass:
      enabled: false
```

## Testing

Verified behavior across scenarios:

- ✅ Fresh deployment with new StorageClasses
- ✅ Existing deployment with pre-existing StorageClasses
- ✅ Mixed environments with some existing, some new StorageClasses
- ✅ Custom StorageClass names and parameters
- ✅ Disabled StorageClass creation
- ✅ Upgrade from previous Helm chart versions

## Files Changed

- `deploy/charts/rook-ceph-cluster/templates/cephblockpool.yaml`
- `deploy/charts/rook-ceph-cluster/templates/cephfilesystem.yaml`
- `deploy/charts/rook-ceph-cluster/templates/cephobjectstore.yaml`
- `deploy/charts/rook-ceph-cluster/templates/_helpers.tpl`

## Related Issues

Resolves the common IaC workflow problem where Rook Helm charts cannot be cleanly applied to existing infrastructure without manual StorageClass management interventions.

dcaputo-harmoni committed:
helm: add automatic StorageClass adoption for Rook Helm charts

Implement intelligent StorageClass detection and adoption to eliminate
conflicts when managing both new and existing Rook deployments in IaC workflows.

- Automatically detect existing StorageClasses using Helm lookup function
- Skip StorageClass rendering when resource already exists to prevent patch conflicts
- Merge user configuration with sensible defaults for minimal configuration
- Apply automatic adoption to CephBlockPool, CephFileSystem, and CephObjectStore
- Maintain full backward compatibility with existing configurations

Fixes StorageClass immutability errors when transitioning from default
configurations to explicit pool specifications in Infrastructure as Code.

Signed-off-by: David Caputo <dcaputo@harmoni.io>

ChristianCiach commented:
I did a quick review on mobile. Looks alright to me! I see nothing that would conflict with non-helm deployments.

dcaputo-harmoni commented:
@travisn just amended my prior commit to include the signoff and also added another to include the functionality for file system and object storage as well.  Let me know if I can be of further help here.

dcaputo-harmoni committed:
helm: add StorageClass parameter helpers for filesystem and object storage

Extends the blockStorageClassParameters helper pattern to filesystem and object storage templates for consistency. This ensures all three storage types use the same approach for generating StorageClass parameters with smart defaults while allowing user customization.

Signed-off-by: David Caputo <dcaputo@harmoni.io>

dcaputo-harmoni commented:
@subhamkrai Thanks for flagging, just fixed - let me know if we are all set.

mkhan-harmoni commented:
I was going to take over from David, but life got in the way
